### PR TITLE
Apply 2457 changes to Timex Java port (2458)

### PR DESCRIPTION
**Pull request by FedericoBernal**

### CODE REVIEW

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/Constants.java [hunk 1]
```diff
@@ -4,6 +4,40 @@
...
+    public static final String MonthUnit = "month";
+    public static final String WeekUnit = "week";
+    public static final String DayUnit = "day";
+    public static final String HourUnit = "hour";
+    public static final String MinuteUnit = "minute";
+    public static final String SecondUnit = "second";
+    public static final String TimeDurationUnit = "s";
+    public static final String AM = "AM";
+    public static final String PM = "PM";
+    public static final int InvalidValue = -1;
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/Constants.java]: Why are we adding all these constants? I mean, the only change in 2457 is TimeNight

Comment (Batta32) [reply Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/Constants.java]: Static props in Java are UPPER_CASE. So TimexNight -> TIMEX_NIGHT

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexConstantsEnglish.java [hunk 2]
```diff
@@ -7,6 +7,15 @@
...
 
 public class TimexConstantsEnglish {
+    public static final String Today = "today";
+    public static final String Tomorrow = "tomorrow";
+    public static final String Yesterday = "yesterday";
+    public static final String Weekend = "weekend";
+    public static final String Tonight = "tonight";
+    public static final String This = "this";
+    public static final String Last = "last";
+    public static final String Next = "next";
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexConstantsEnglish.java]: UPPER_CASE

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexConvertEnglish.java [hunk 3]
```diff
@@ -86,17 +86,20 @@ public static String convertDate(TimexProperty timex) {
...
         String abbreviation = TimexConstantsEnglish.DATE_ABBREVIATION[Integer
                 .parseInt(String.valueOf(date.charAt(date.length() - 1)))];
 
-        if (timex.getYear() != null) {
-            return String.format("%1$s%2$s %3$s %4$s", date, abbreviation, month, timex.getYear()).trim();
+        if (timex.getMonth() != null) {
+            String month = TimexConstantsEnglish.MONTHS[timex.getMonth() - 1];
+            if (timex.getYear() != null)
+            {
+                return String.format(date, abbreviation, month, timex.getYear());
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexConvertEnglish.java]: This is not the same as [C#](https://github.com/microsoft/Recognizers-Text/pull/2457/files#diff-246366d7220774e90623ebc0472c32bc5f3cb025976aa994ef12cae0e023b976R107).

Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexConvertEnglish.java]:
```suggestion
                return String.format("%1$s%2$s %3$s %4$s", date, abbreviation, month, timex.getYear()).trim();
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexConvertEnglish.java [hunk 4]
```diff
@@ -86,17 +86,20 @@ public static String convertDate(TimexProperty timex) {
...
+            }
+            return String.format(date, abbreviation, month);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexConvertEnglish.java]:
```suggestion
            return String.format("%1$s%2$s %3$s", date, abbreviation, month);
```

Following [C#](https://github.com/microsoft/Recognizers-Text/pull/2457/files#diff-246366d7220774e90623ebc0472c32bc5f3cb025976aa994ef12cae0e023b976R110)

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexConvertEnglish.java [hunk 5]
```diff
@@ -86,17 +86,20 @@ public static String convertDate(TimexProperty timex) {
...
         }
-
-        return String.format("%1$s%2$s %3$s", date, abbreviation, month);
+        return String.format(date, abbreviation);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexConvertEnglish.java]:
```suggestion
        return date.concat(abbreviation);
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 6]
```diff
@@ -45,31 +45,31 @@ private static String convertDate(TimexProperty timex, LocalDateTime date) {
...
             LocalDateTime yesterday = TimexDateHelpers.yesterday(date);
             if (TimexDateHelpers.datePartEquals(timexDate, yesterday)) {
-                return "yesterday";
+                return TimexConstantsEnglish.Yesterday;
             }
 
             if (TimexDateHelpers.isThisWeek(timexDate, date)) {
-                return String.format("this %s",
+                return String.format(TimexConstantsEnglish.This,
                         TimexRelativeConvertEnglish.getDateDay(timexDate.getDayOfWeek()));
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                return String.format("%1$s %2$s", TimexConstantsEnglish.This,
                        TimexRelativeConvertEnglish.getDateDay(timexDate.getDayOfWeek()));
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 7]
```diff
@@ -45,31 +45,31 @@ private static String convertDate(TimexProperty timex, LocalDateTime date) {
...
             }
 
             if (TimexDateHelpers.isNextWeek(timexDate, date)) {
-                return String.format("next %s",
+                return String.format(TimexConstantsEnglish.Next,
                         TimexRelativeConvertEnglish.getDateDay(timexDate.getDayOfWeek()));
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                return String.format("%1$s %2$s", TimexConstantsEnglish.Next,
                        TimexRelativeConvertEnglish.getDateDay(timexDate.getDayOfWeek()));
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 8]
```diff
@@ -45,31 +45,31 @@ private static String convertDate(TimexProperty timex, LocalDateTime date) {
...
             }
 
             if (TimexDateHelpers.isLastWeek(timexDate, date)) {
-                return String.format("last %s",
+                return String.format(TimexConstantsEnglish.Last,
                         TimexRelativeConvertEnglish.getDateDay(timexDate.getDayOfWeek()));
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                return String.format("%1$s %2$s", TimexConstantsEnglish.Last,
                        TimexRelativeConvertEnglish.getDateDay(timexDate.getDayOfWeek()));
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 9]
```diff
@@ -89,45 +89,45 @@ private static String convertDateRange(TimexProperty timex, LocalDateTime date)
                 if (timex.getWeekOfYear() != null) {
                     Integer thisWeek = TimexDateHelpers.weekOfYear(date);
                     if (thisWeek == timex.getWeekOfYear()) {
-                        return timex.getWeekend() != null ? "this weekend" : "this week";
+                        return timex.getWeekend() != null ? String.format(TimexConstantsEnglish.This, TimexConstantsEnglish.Weekend) : String.format(TimexConstantsEnglish.This, Constants.WeekUnit);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                        return timex.getWeekend() != null ? String.format("%1$s %2$s", TimexConstantsEnglish.This, TimexConstantsEnglish.Weekend) : String.format("%1$s %2$s", TimexConstantsEnglish.This, Constants.WeekUnit);
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 10]
```diff
@@ -89,45 +89,45 @@ private static String convertDateRange(TimexProperty timex, LocalDateTime date)
...
                     }
 
                     if (thisWeek == timex.getWeekOfYear() + 1) {
-                        return timex.getWeekend() != null ? "last weekend" : "last week";
+                        return timex.getWeekend() != null ? String.format(TimexConstantsEnglish.This, TimexConstantsEnglish.Weekend) : String.format(TimexConstantsEnglish.Last, Constants.WeekUnit);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                        return timex.getWeekend() != null ? String.format("%1$s %2$s", TimexConstantsEnglish.Last, TimexConstantsEnglish.Weekend) : String.format("%1$s %2$s", TimexConstantsEnglish.Last, Constants.WeekUnit);
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 11]
```diff
@@ -89,45 +89,45 @@ private static String convertDateRange(TimexProperty timex, LocalDateTime date)
...
                     }
 
                     if (thisWeek == timex.getWeekOfYear() - 1) {
-                        return timex.getWeekend() != null ? "next weekend" : "next week";
+                        return timex.getWeekend() != null ? String.format(TimexConstantsEnglish.This, TimexConstantsEnglish.Weekend) : String.format(TimexConstantsEnglish.Next, Constants.WeekUnit);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                        return timex.getWeekend() != null ? String.format("%1$s %2$s", TimexConstantsEnglish.Next, TimexConstantsEnglish.Weekend) : String.format("%1$s %2$s", TimexConstantsEnglish.Next, Constants.WeekUnit);
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 12]
```diff
@@ -89,45 +89,45 @@ private static String convertDateRange(TimexProperty timex, LocalDateTime date)
...
                     }
                 }
 
                 if (timex.getMonth() != null) {
                     if (timex.getMonth() == date.getMonthValue()) {
-                        return "this month";
+                        return String.format(TimexConstantsEnglish.This, Constants.MonthUnit);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                        return String.format("%1$s %2$s", TimexConstantsEnglish.This, Constants.MonthUnit);
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 13]
```diff
@@ -89,45 +89,45 @@ private static String convertDateRange(TimexProperty timex, LocalDateTime date)
...
                     }
 
                     if (timex.getMonth() == date.getMonthValue() + 1) {
-                        return "next month";
+                        return String.format(TimexConstantsEnglish.Next, Constants.MonthUnit);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                        return String.format("%1$s %2$s", TimexConstantsEnglish.Next, Constants.MonthUnit);
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 14]
```diff
@@ -89,45 +89,45 @@ private static String convertDateRange(TimexProperty timex, LocalDateTime date)
...
                     }
 
                     if (timex.getMonth() == date.getMonthValue() - 1) {
-                        return "last month";
+                        return String.format(TimexConstantsEnglish.Last, Constants.MonthUnit);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                        return String.format("%1$s %2$s", TimexConstantsEnglish.Last, Constants.MonthUnit);
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 15]
```diff
@@ -89,45 +89,45 @@ private static String convertDateRange(TimexProperty timex, LocalDateTime date)
...
                     }
                 }
 
 
-                return (timex.getSeason() != null) ? String.format("this %s", TimexConstantsEnglish.SEASONS.get(timex.getSeason()))
-                            : "this year";
+                return (timex.getSeason() != null) ? String.format(TimexConstantsEnglish.This, TimexConstantsEnglish.SEASONS.get(timex.getSeason()))
+                            : String.format(TimexConstantsEnglish.This, Constants.YearUnit);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                return (timex.getSeason() != null) ? String.format("%1$s %2$s", TimexConstantsEnglish.This, TimexConstantsEnglish.SEASONS.get(timex.getSeason()))
                            : String.format("%1$s %2$s", TimexConstantsEnglish.This, Constants.YearUnit);
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 16]
```diff
@@ -89,45 +89,45 @@ private static String convertDateRange(TimexProperty timex, LocalDateTime date)
...
             }
 
             if (timex.getYear() == year + 1) {
-                return (timex.getSeason() != null) ? String.format("next %s", TimexConstantsEnglish.SEASONS.get(timex.getSeason()))
-                    : "next year";
+                return (timex.getSeason() != null) ? String.format(TimexConstantsEnglish.Next, TimexConstantsEnglish.SEASONS.get(timex.getSeason()))
+                    : String.format(TimexConstantsEnglish.Next, Constants.YearUnit);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                return (timex.getSeason() != null) ? String.format("%1$s %2$s", TimexConstantsEnglish.Next, TimexConstantsEnglish.SEASONS.get(timex.getSeason()))
                    : String.format("%1$s %2$s", TimexConstantsEnglish.Next, Constants.YearUnit);
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 17]
```diff
@@ -89,45 +89,45 @@ private static String convertDateRange(TimexProperty timex, LocalDateTime date)
...
             }
 
             if (timex.getYear() == year - 1) {
-                return (timex.getSeason() != null) ? String.format("last %s", TimexConstantsEnglish.SEASONS.get(timex.getSeason()))
-                    : "last year";
+                return (timex.getSeason() != null) ? String.format(TimexConstantsEnglish.Last, TimexConstantsEnglish.SEASONS.get(timex.getSeason()))
+                    : String.format(TimexConstantsEnglish.Last, Constants.YearUnit);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                return (timex.getSeason() != null) ? String.format("%1$s %2$s", TimexConstantsEnglish.Last, TimexConstantsEnglish.SEASONS.get(timex.getSeason()))
                    : String.format("%1$s %2$s", TimexConstantsEnglish.Last, Constants.YearUnit);
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 18]
```diff
@@ -140,31 +140,31 @@ private static String convertDateTimeRange(TimexProperty timex, LocalDateTime da
 
             if (timex.getPartOfDay() != null) {
                 if (TimexDateHelpers.datePartEquals(timexDate, date)) {
-                    if (timex.getPartOfDay().equals("NI")) {
-                        return "tonight";
+                    if (timex.getPartOfDay().equals(Constants.TimexNight)) {
+                        return TimexConstantsEnglish.Tonight;
                     } else {
-                        return String.format("this %s", TimexConstantsEnglish.DAY_PARTS.get(timex.getPartOfDay()));
+                        return String.format(TimexConstantsEnglish.This, TimexConstantsEnglish.DAY_PARTS.get(timex.getPartOfDay()));
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                        return String.format"%1$s %2$s", TimexConstantsEnglish.This, TimexConstantsEnglish.DAY_PARTS.get(timex.getPartOfDay()));
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 19]
```diff
@@ -140,31 +140,31 @@ private static String convertDateTimeRange(TimexProperty timex, LocalDateTime da
...
                     }
                 }
 
                 LocalDateTime tomorrow = TimexDateHelpers.tomorrow(date);
                 if (TimexDateHelpers.datePartEquals(timexDate, tomorrow)) {
-                    return String.format("tomorrow %s", TimexConstantsEnglish.DAY_PARTS.get(timex.getPartOfDay()));
+                    return String.format(TimexConstantsEnglish.Tomorrow, TimexConstantsEnglish.DAY_PARTS.get(timex.getPartOfDay()));
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                    return String.format("%1$s %2$s", TimexConstantsEnglish.Tomorrow, TimexConstantsEnglish.DAY_PARTS.get(timex.getPartOfDay()));
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 20]
```diff
@@ -140,31 +140,31 @@ private static String convertDateTimeRange(TimexProperty timex, LocalDateTime da
...
                 }
 
                 LocalDateTime yesterday = TimexDateHelpers.yesterday(date);
                 if (TimexDateHelpers.datePartEquals(timexDate, yesterday)) {
-                    return String.format("yesterday %s", TimexConstantsEnglish.DAY_PARTS.get(timex.getPartOfDay()));
+                    return String.format(TimexConstantsEnglish.Yesterday, TimexConstantsEnglish.DAY_PARTS.get(timex.getPartOfDay()));
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                    return String.format("%1$s %2$s", TimexConstantsEnglish.Yesterday, TimexConstantsEnglish.DAY_PARTS.get(timex.getPartOfDay()));
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 21]
```diff
@@ -140,31 +140,31 @@ private static String convertDateTimeRange(TimexProperty timex, LocalDateTime da
...
                 }
 
                 if (TimexDateHelpers.isNextWeek(timexDate, date)) {
-                    return String.format("next %1$s %2$s",
+                    return String.format(TimexConstantsEnglish.Next,
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                    return String.format("%1$s %2$s", TimexConstantsEnglish.Next,
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java [hunk 22]
```diff
@@ -140,31 +140,31 @@ private static String convertDateTimeRange(TimexProperty timex, LocalDateTime da
...
                             TimexRelativeConvertEnglish.getDateDay(timexDate.getDayOfWeek()),
                             TimexConstantsEnglish.DAY_PARTS.get(timex.getPartOfDay()));
                 }
 
                 if (TimexDateHelpers.isLastWeek(timexDate, date)) {
-                    return String.format("last %1$s %2$s",
+                    return String.format(TimexConstantsEnglish.Last,
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexRelativeConvertEnglish.java]:
```suggestion
                    return String.format("%1$s %2$s", TimexConstantsEnglish.Last,
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexFormat.java [hunk 23]
```diff
@@ -111,7 +111,8 @@ private static String formatTime(TimexProperty timex) {
     private static String formatDate(TimexProperty timex) {
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexFormat.java:123]: All this method was reduced to one line in [C#](https://github.com/microsoft/Recognizers-Text/pull/2457/files#diff-4f811f041fb2b829ff5ec51e6b8a43862c8cbcb3a7552965a33b97e1aa561dedR124).

Why do we have all these here?

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexFormat.java [hunk 24]
```diff
@@ -138,6 +139,11 @@ private static String formatDateRange(TimexProperty timex) {
                     TimexDateHelpers.fixedFormatNumber(timex.getWeekOfYear(), 2));
         }
 
+        if (timex.getYear() != null && timex.getMonth() != null && timex.getWeekOfMonth() != null)
+        {
+            return String.format(TimexDateHelpers.fixedFormatNumber(timex.getYear(), 4), TimexDateHelpers.fixedFormatNumber(timex.getMonth(), 2), TimexDateHelpers.fixedFormatNumber(timex.getWeekOfMonth(), 2));
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexFormat.java]:
```suggestion
            return String.format("%1$s-%2$s-W%3$s", TimexDateHelpers.fixedFormatNumber(timex.getYear(), 4), TimexDateHelpers.fixedFormatNumber(timex.getMonth(), 2), TimexDateHelpers.fixedFormatNumber(timex.getWeekOfMonth(), 2));
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java [hunk 25]
```diff
@@ -178,6 +179,31 @@ public static TimexProperty timexDateAdd(TimexProperty start, TimexProperty dura
...
     }
 
+    public static String generateDateTimex(Integer year, Integer monthOrWeekOfYear, Integer day, Integer weekOfMonth, boolean byWeek)
+    {
+        String yearString = year == Constants.InvalidValue ? Constants.TimexFuzzyYear : TimexDateHelpers.fixedFormatNumber(year, 4);
+        String monthWeekString = monthOrWeekOfYear == Constants.InvalidValue ? Constants.TimexFuzzyMonth : TimexDateHelpers.fixedFormatNumber(monthOrWeekOfYear, 2);
+        String dayString;
+        if (byWeek)
+        {
+            dayString = day.toString(Integer.parseInt(Locale.ROOT.getLanguage()));
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java]: Not quite sure about this: `toString(Integer.parseInt(Locale.ROOT.getLanguage()))`

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java [hunk 26]
```diff
@@ -178,6 +179,31 @@ public static TimexProperty timexDateAdd(TimexProperty start, TimexProperty dura
...
+            if (weekOfMonth != Constants.InvalidValue)
+            {
+                monthWeekString = monthWeekString + String.format(Constants.TimexFuzzyWeek, weekOfMonth.toString(Integer.parseInt(Locale.ROOT.getLanguage())));
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java]:
```suggestion
                monthWeekString = monthWeekString + String.format(-%s-, Constants.TimexFuzzyWeek) + weekOfMonth.toString(Integer.parseInt(Locale.ROOT.getLanguage()));
```

Comment (Batta32) [reply Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java]: Also not sure about `toString(Integer.parseInt(Locale.ROOT.getLanguage()))`

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java [hunk 27]
```diff
@@ -178,6 +179,31 @@ public static TimexProperty timexDateAdd(TimexProperty start, TimexProperty dura
...
+            {
+                monthWeekString = Constants.TimexWeek + monthWeekString;
+            }
+        }
+        else
+        {
+            dayString = day == Constants.InvalidValue ? Constants.TimexFuzzyDay : TimexDateHelpers.fixedFormatNumber(day, 2);
+        }
+
+        return String.format(yearString, monthWeekString, dayString);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java]:
```suggestion
        return String.join("-", yearString, monthWeekString, dayString);
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java [hunk 28]
```diff
@@ -265,6 +291,10 @@ public static TimeRange timeRangeFromTimex(TimexProperty timex) {
             }
         };
     }
+    public static String formatResolvedDateValue(String dateValue, String timeValue)
+    {
+        return String.format(dateValue, timeValue);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java]:
```suggestion
        return String.format("%1$s %2$s", dateValue, timeValue);
```

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexProperty.java [hunk 29]
```diff
@@ -408,6 +408,7 @@ public void assignProperties(Map<String, String> source) {
                     this.assignTimeDuration(source);
                     break;
                 default:
+                    continue;
```
Comment (Batta32) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexProperty.java]: [C#](https://github.com/microsoft/Recognizers-Text/pull/2457/files#diff-30c57aa18922d5b3781bb097653c619215ad8ab665f8dec20d89ce8cabafec1fR203) checks if the value is empty so the `switch` is not executed. Here we are not doing this.

**Review by Batta32 — CHANGES_REQUESTED** [review]

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java (hunk 26 above)
Comment (matiasroldan6) [reply Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java]: If I'm not mistaken, the toString isn't even necessary on the weekOfMonth. The String.format method takes an array of Object

Comment (Batta32) [reply Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java]: If that's the case, lgtm

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexDateHelpers.java [hunk 30]
```diff
@@ -119,4 +119,8 @@ public static LocalDateTime dateOfNextDay(DayOfWeek day, LocalDateTime reference
 
         return result;
     }
+
+    public static Integer getUSDateOfWeekValue(DayOfWeek dayOfWeek) {
```
Comment (matiasroldan6) [inline Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexDateHelpers.java]: I'm not seeing these changes in the PR 2457, or in the C# file in that branch, but I assume this was necessary

Comment (Batta32) [reply Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexDateHelpers.java]: This is a disparity between how DayOfWeek (which I noticed that this should be renamed to `getUSDayOfweek`) works in C# vs Java.

C# has Sunday as 0
Java has Sunday as 7

This was necessary to have a parity

**Review by Batta32 — APPROVED** [review]
LGTM

**Pull request merged**